lib/dbvar/test/unit/Data/DeltaSpec.hs

```diff
@@ -1,18 +1,78 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
 module Data.DeltaSpec
     ( spec
     ) where
 
 import Prelude
 
+import Data.DBVar
+    ( Store (..), newCachedStore, newStore )
+import Data.Delta
+import Fmt
+import Test.DBVar
+    ( prop_StoreUpdates )
 import Test.Hspec
     ( Spec, describe, it, parallel )
+import Test.QuickCheck
+    ( elements, generate, (===) )
+import Test.QuickCheck.Gen
+    ( Gen, listOf )
+import Test.QuickCheck.Monadic
+    ( monadicIO, run )
 
 spec :: Spec
 spec = do
     parallel $ describe "Data.Delta" $ do
         it "Dummy test, to be expanded"
             True
+    describe "CachedStore" $ do
+        it "respects store laws" $ monadicIO $ do
+            cachedStore <- run $ do
+                testStore <- newStore @TestStoreDelta
+                writeS testStore 0
+                newCachedStore testStore
+            prop_StoreUpdates run
+                cachedStore
+                (pure 0)
+                $ const genTestStoreDeltas
+        it "behaves like the cached one" $ monadicIO $ do
+            das <- run $ generate $ listOf genTestStoreDeltas
+            (cachedFinal, originalFinal)<- run $ do
+                testStore <- newStore @TestStoreDelta
+                writeS testStore 0
+                cachedStore <- newCachedStore testStore
+                updateStore cachedStore das
+                Right cachedFinal <- loadS cachedStore
+                writeS testStore 0
+                updateStore testStore das
+                Right originalFinal <- loadS testStore
+                pure (cachedFinal, originalFinal)
+            pure $ cachedFinal === originalFinal
+
+updateStore :: Monad m => Store m da -> [da] -> m ()
+updateStore store = mapM_ (updateS store Nothing)
+
+genTestStoreDeltas :: Gen TestStoreDelta
+genTestStoreDeltas = elements [AddOne, AddTwo, RemoveOne]
+
+newtype TestStoreBase = TestStoreBase Int
+    deriving (Show, Eq, Num )
+
+data TestStoreDelta
+    = AddOne
+    | AddTwo
+    | RemoveOne
+
+    deriving (Show, Eq)
+instance Buildable TestStoreDelta where
+    build = build . show
 
-{-------------------------------------------------------------------------------
-    Properties
--------------------------------------------------------------------------------}
+instance Delta TestStoreDelta where
+    type Base TestStoreDelta = TestStoreBase
+    apply AddOne = ( + 1)
+    apply AddTwo = ( + 2)
+    apply RemoveOne = ( subtract 1)
```

Comment:
```suggestion
    apply AddOne = overTestStoreBase (1:)
    apply AddTwo = overTestStoreBase (2:)
    apply RemoveOne = overTestStoreBase (drop 1 xs)

overTestStoreBase :: ([Int] -> [Int]) -> TestStoreBase -> TestStoreBase
overTestStoreBase f (TestStoreBase xs) = TestStoreBase (f xs)
```

Slightly more verbose, but using a list `[Int]` instead of plain `Int` here has the advantage that the deltas are not commutative, that is we have

```
apply AddTwo . apply AddOne  =/=  apply AddOne . apply AddTwo
```

This could, in principle, reveal more bugs.